src/ripple/rpc/handlers/LedgerData.cpp

```diff
@@ -78,28 +79,29 @@ Json::Value doLedgerData (RPC::Context& context)
     jvResult[jss::ledger_hash] = to_string (lpLedger->info().hash);
     jvResult[jss::ledger_index] = std::to_string(lpLedger->info().seq);
 
-    Json::Value& nodes = (jvResult[jss::state] = Json::arrayValue);
+    Json::Value& nodes = jvResult[jss::state];
 
-    for(auto const& sle : lpLedger->sles)
+    while ((key = lpLedger->succ(*key)))
     {
-       if (limit-- <= 0)
-       {
-           auto marker = sle->key();
-           jvResult[jss::marker] = to_string (--marker);
-           break;
-       }
-
-       if (isBinary)
-       {
-           Json::Value& entry = nodes.append (Json::objectValue);
-           entry[jss::data] = serializeHex(*sle);
-           entry[jss::index] = to_string (sle->key());
-       }
-       else
-       {
-           Json::Value& entry = nodes.append (sle->getJson (0));
-           entry[jss::index] = to_string (sle->key());
-       }
+        auto sle = lpLedger->read(keylet::unchecked(*key));
+        if (limit-- <= 0)
+        {
```

Comment:
> If we don't hit the limit we can just use the key of the last entry as
the marker

I think It's more like one just isn't needed.

On the client side, you know you don't need to make another request and you
have all the items by the absence of the marker variable.